File: LeetCode/205IsomorphicStrings/IsomorphicStrings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        // 2 dictionaries -> Time:O(n), Space:O(n)
        map<char, char> s_map;
        map<char, char> t_map;
        for (int i = 0; i < s.size(); i++) {
            if (s_map.count(s[i]) == 0 && t_map.count(t[i]) == 0) {
                s_map[s[i]] = t[i];
                t_map[t[i]] = s[i];
            }
            else if (s_map.count(s[i]) == 0 || t_map.count(t[i]) == 0) {
                return false;
            }
            else if (s_map[s[i]] != t[i] || t_map[t[i]] != s[i]) {
                return false;
            }
        }
        return true;
    }
};

File: LeetCode/205IsomorphicStrings/IsomorphicStrings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        # 2 dictionaries -> Time:O(n), Space:O(n)
        s_d = {}
        t_d = {}
        for i in range(len(s)):
            if s[i] not in s_d and t[i] not in t_d:
                s_d[s[i]] = t[i]
                t_d[t[i]] = s[i]
            elif s[i] not in s_d or t[i] not in t_d:
                return False
            elif s_d[s[i]] != t[i] or t_d[t[i]] != s[i]:
                return False
        return True
